File: lib/herb/range.rb
# frozen_string_literal: true

module Herb
  class Range
    attr_reader :from, :to

    def initialize(from, to)
      @from = from
      @to = to
    end

    def self.[](...)
      new(...)
    end

    def self.from(...)
      new(...)
    end

    def to_a
      [from, to]
    end

    def to_json(*args)
      to_a.to_json(*args)
    end

    def tree_inspect
      to_a.to_s
    end

    def inspect
      %(#<Herb::Range #{to_a}>)
    end

    def to_s
      inspect
    end
  end
end

File: javascript/packages/core/src/range.ts
export type SerializedRange = [number, number]

export class Range {
  readonly start: number
  readonly end: number

  static from(range: SerializedRange): Range
  static from(start: number, end: number): Range
  static from(rangeOrStart: SerializedRange | number, end?: number): Range {
    if (typeof rangeOrStart === "number") {
      return new Range(rangeOrStart, end!)
    } else {
      return new Range(rangeOrStart[0], rangeOrStart[1])
    }
  }

  static get zero() {
    return new Range(0, 0)
  }

  constructor(start: number, end: number) {
    this.start = start
    this.end = end
  }

  toArray(): SerializedRange {
    return [this.start, this.end]
  }

  toJSON(): SerializedRange {
    return this.toArray()
  }

  treeInspect(): string {
    return `[${this.start}, ${this.end}]`
  }

  inspect(): string {
    return `#<Herb::Range ${this.toArray()}>`
  }

  toString(): string {
    return this.inspect()
  }
}
